### PR TITLE
Add MIR Optimization Tests

### PR DESCRIPTION
**Pull request by scottcarr**
I've starting working on the infrastructure for testing MIR optimizations.

The plan now is to have a set of test cases (written in Rust), compile them with -Z dump-mir, and check the MIR before and after each pass.

### CODE REVIEW

File: src/librustc_mir/pretty.rs [hunk 1]
```diff
@@ -66,9 +66,15 @@ pub fn dump_mir<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
...
     };
 
+    let mut file_path = String::new();
+    if let Some(ref file_dir) = tcx.sess.opts.debugging_opts.dump_mir_dir {
+        file_path.push_str(file_dir);
+    };
     let file_name = format!("rustc.node{}{}.{}.{}.mir",
                             node_id, promotion_id, pass_name, disambiguator);
-    let _ = fs::File::create(&file_name).and_then(|mut file| {
+    file_path.push_str("/");
```
Comment (rkruppe) [inline src/librustc_mir/pretty.rs]: Doesn't this mean that, if no path is specified, the MIR will be dumped into the root folter (`/rustc.node[...].mir`)? Using `PathBuf` can avoid that, and would be semantically more appropriate and more robust anyway.

**Comment by alexcrichton** [comment]
Makefile changes look good to me! Could you also be sure to update `src/bootstrap/step.rs` to ensure that rustbuild runs this compiletest test suite as well?

**Comment by scottcarr** [comment]
@alexcrichton is there a bot that runs rustbuild?  I think the travis bot builds with make, right?  I'm pretty sure I tested this with both build systems, but we could double check.

**Comment by alexcrichton** [comment]
@scottcarr all correct! I'm fine letting bors figure out the rest

File: src/tools/compiletest/src/runtest.rs [hunk 2]
```diff
@@ -1336,7 +1337,24 @@ actual:\n\
...
+                args.extend(["-Z",
+                             "dump-mir=all",
+                             "-Z"]
+                            .iter()
+                            .map(|s| s.to_string()));
 
+
+                let mir_dump_dir = self.get_mir_dump_dir();
+                if !mir_dump_dir.exists() {
+                    fs::create_dir(mir_dump_dir.clone()).expect("the dir should exist");
```
Comment (nikomatsakis) [inline src/tools/compiletest/src/runtest.rs]: I think this should be the "racy" create-dir option, probably-- though I guess each test will be making disjoint directories, so maybe it's not needed.

**Comment by nikomatsakis** [comment]
Can you maybe add the `src/test/mir-dump` directory (or whatever it is to be called) along with a `README.md` that just (briefly) explains how such a test should look?

**Comment by nikomatsakis** [comment]
But basically r+ from me

**Comment by alexcrichton** [comment]
@bors: retry

* #34595

**Pull request merged**